File: go/test/repo_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/github-sdk/go"
	"github.com/voxgig-sdk/github-sdk/go/core"

	vs "github.com/voxgig-sdk/github-sdk/go/utility/struct"
)

func TestRepoEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Repo(nil)
		if ent == nil {
			t.Fatal("expected non-nil RepoEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"repo": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Repo(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.SharedConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Repo(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := repoBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "repo." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set GITHUB_TEST_REPO_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		repoRef01Ent := client.Repo(nil)
		repoRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "repo"}), "repo_ref01"))
		repoRef01Data["branch_id"] = setup.idmap["branch01"]
		repoRef01Data["environment_id"] = setup.idmap["environment01"]
		repoRef01Data["environment_name"] = setup.idmap["environment_name01"]
		repoRef01Data["org_id"] = setup.idmap["org01"]
		repoRef01Data["owner"] = setup.idmap["owner01"]
		repoRef01Data["repo"] = setup.idmap["repo01"]

		repoRef01DataResult, err := repoRef01Ent.Create(repoRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		repoRef01Data = core.ToMapAny(entityData(repoRef01DataResult))
		if repoRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if repoRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		repoRef01Match := map[string]any{
			"environment_name": setup.idmap["environment_name01"],
			"owner": setup.idmap["owner01"],
			"repo": setup.idmap["repo01"],
		}

		repoRef01ListResult, err := repoRef01Ent.List(repoRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		repoRef01List, repoRef01ListOk := repoRef01ListResult.([]any)
		if !repoRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", repoRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(repoRef01List), map[string]any{"id": repoRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		repoRef01DataUp0Up := map[string]any{
			"id": repoRef01Data["id"],
			"owner": setup.idmap["owner"],
		}

		repoRef01MarkdefUp0Name := "archive_url"
		repoRef01MarkdefUp0Value := fmt.Sprintf("Mark01-repo_ref01_%d", setup.now)
		repoRef01DataUp0Up[repoRef01MarkdefUp0Name] = repoRef01MarkdefUp0Value

		repoRef01ResdataUp0Result, err := repoRef01Ent.Update(repoRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		repoRef01ResdataUp0 := core.ToMapAny(entityData(repoRef01ResdataUp0Result))
		if repoRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if repoRef01ResdataUp0["id"] != repoRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if repoRef01ResdataUp0[repoRef01MarkdefUp0Name] != repoRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", repoRef01MarkdefUp0Name, repoRef01ResdataUp0[repoRef01MarkdefUp0Name])
		}

		// LOAD
		repoRef01MatchDt0 := map[string]any{
			"id": repoRef01Data["id"],
		}
		repoRef01DataDt0Loaded, err := repoRef01Ent.Load(repoRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		repoRef01DataDt0LoadResult := core.ToMapAny(entityData(repoRef01DataDt0Loaded))
		if repoRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if repoRef01DataDt0LoadResult["id"] != repoRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		repoRef01MatchRm0 := map[string]any{
			"id": repoRef01Data["id"],
		}
		_, err = repoRef01Ent.Remove(repoRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		repoRef01MatchRt0 := map[string]any{
			"environment_name": setup.idmap["environment_name01"],
			"owner": setup.idmap["owner01"],
			"repo": setup.idmap["repo01"],
		}

		repoRef01ListRt0Result, err := repoRef01Ent.List(repoRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		repoRef01ListRt0, repoRef01ListRt0Ok := repoRef01ListRt0Result.([]any)
		if !repoRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", repoRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(repoRef01ListRt0), map[string]any{"id": repoRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func repoBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "repo", "RepoTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read repo test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse repo test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"repo01", "repo02", "repo03", "repository_invitation01", "repository_invitation02", "repository_invitation03", "org01", "org02", "org03", "ruleset01", "ruleset02", "ruleset03", "attestation01", "attestation02", "attestation03", "autolink01", "autolink02", "autolink03", "branch01", "branch02", "branch03", "collaborator01", "collaborator02", "collaborator03", "comment01", "comment02", "comment03", "content01", "content02", "content03", "deployment01", "deployment02", "deployment03", "environment01", "environment02", "environment03", "hook01", "hook02", "hook03", "invitation01", "invitation02", "invitation03", "key01", "key02", "key03", "asset01", "asset02", "asset03", "release01", "release02", "release03", "protection01", "protection02", "protection03", "tarball01", "tarball02", "tarball03", "zipball01", "zipball02", "zipball03", "deployment_branch_policy01", "deployment_branch_policy02", "deployment_branch_policy03", "deployment_protection_rule01", "deployment_protection_rule02", "deployment_protection_rule03", "delivery01", "delivery02", "delivery03", "environment_name01", "owner01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("GITHUB_TEST_REPO_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"GITHUB_TEST_REPO_ENTID": idmap,
		"GITHUB_TEST_LIVE":      "FALSE",
		"GITHUB_TEST_EXPLAIN":   "FALSE",
		"GITHUB_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["GITHUB_TEST_REPO_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add owner alias for update test.
	if idmapResolved["owner"] == nil {
		idmapResolved["owner"] = idmapResolved["owner01"]
	}

	if env["GITHUB_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["GITHUB_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewGithubSDK(core.ToMapAny(mergedOpts))
	}

	live := env["GITHUB_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["GITHUB_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: lua/test/repo_entity_test.lua
-- Repo entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("github_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("RepoEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:Repo(nil)
    assert.is_not_nil(ent)
  end)

  -- Feature #4: the entity stream(action, ...) method runs the op pipeline and
  -- returns an iterator over result items. With the streaming feature active it
  -- yields the feature's incremental output; otherwise it falls back to the
  -- materialised list so stream always yields.
  it("should stream", function()
    local seed = {
      entity = {
        ["repo"] = {
          s1 = { id = "s1" },
          s2 = { id = "s2" },
          s3 = { id = "s3" },
        },
      },
    }

    -- Fallback: streaming inactive -> yields the materialised list items.
    local base = sdk.test(seed, nil)
    local seen = {}
    for item in base:Repo(nil):stream("list", nil, nil) do
      table.insert(seen, item)
    end
    assert.are.equal(3, #seen)

    -- Inbound: streaming active -> yields each item from the feature.
    local config = require("config_shared")()
    if type(config.feature) == "table" and config.feature.streaming ~= nil then
      local streamsdk = sdk.test(seed, { feature = { streaming = { active = true } } })
      local got = {}
      for item in streamsdk:Repo(nil):stream("list", nil, nil) do
        if vs.islist(item) then
          for _, sub in ipairs(item) do
            table.insert(got, sub)
          end
        else
          table.insert(got, item)
        end
      end
      assert.are.equal(3, #got)
    end
  end)

  it("should run basic flow", function()
    local setup = repo_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create", "list", "update", "load", "remove"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "repo." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set GITHUB_TEST_REPO_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local repo_ref01_ent = client:Repo(nil)
    local repo_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.repo"), "repo_ref01"))
    repo_ref01_data["owner"] = setup.idmap["owner01"]

    local repo_ref01_data_result, err = repo_ref01_ent:create(repo_ref01_data, nil)
    assert.is_nil(err)
    repo_ref01_data = helpers.to_map(type(repo_ref01_data_result) == 'table' and repo_ref01_data_result.data_get and repo_ref01_data_result:data_get() or repo_ref01_data_result)
    assert.is_not_nil(repo_ref01_data)
    assert.is_not_nil(repo_ref01_data["id"])

    -- LIST
    local repo_ref01_match = {}

    local repo_ref01_list_result, err = repo_ref01_ent:list(repo_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(repo_ref01_list_result)

    local found_item = vs.select(
      runner.entity_list_to_data(repo_ref01_list_result),
      { id = repo_ref01_data["id"] })
    assert.is_false(vs.isempty(found_item))

    -- UPDATE
    local repo_ref01_data_up0_up = {
      id = repo_ref01_data["id"],
      ["owner"] = setup.idmap["owner"],
    }

    local repo_ref01_markdef_up0_name = "archive_url"
    local repo_ref01_markdef_up0_value = "Mark01-repo_ref01_" .. tostring(setup.now)
    repo_ref01_data_up0_up[repo_ref01_markdef_up0_name] = repo_ref01_markdef_up0_value

    local repo_ref01_resdata_up0_result, err = repo_ref01_ent:update(repo_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local repo_ref01_resdata_up0 = helpers.to_map(type(repo_ref01_resdata_up0_result) == 'table' and repo_ref01_resdata_up0_result.data_get and repo_ref01_resdata_up0_result:data_get() or repo_ref01_resdata_up0_result)
    assert.is_not_nil(repo_ref01_resdata_up0)
    assert.are.equal(repo_ref01_resdata_up0["id"], repo_ref01_data_up0_up["id"])
    assert.are.equal(repo_ref01_resdata_up0[repo_ref01_markdef_up0_name], repo_ref01_markdef_up0_value)

    -- LOAD
    local repo_ref01_match_dt0 = {
      id = repo_ref01_data["id"],
    }
    local repo_ref01_data_dt0_loaded, err = repo_ref01_ent:load(repo_ref01_match_dt0, nil)
    assert.is_nil(err)
    local repo_ref01_data_dt0_load_result = helpers.to_map(type(repo_ref01_data_dt0_loaded) == 'table' and repo_ref01_data_dt0_loaded.data_get and repo_ref01_data_dt0_loaded:data_get() or repo_ref01_data_dt0_loaded)
    assert.is_not_nil(repo_ref01_data_dt0_load_result)
    assert.are.equal(repo_ref01_data_dt0_load_result["id"], repo_ref01_data["id"])

    -- REMOVE
    local repo_ref01_match_rm0 = {
      id = repo_ref01_data["id"],
    }
    local _, err = repo_ref01_ent:remove(repo_ref01_match_rm0, nil)
    assert.is_nil(err)

    -- LIST
    local repo_ref01_match_rt0 = {}

    local repo_ref01_list_rt0_result, err = repo_ref01_ent:list(repo_ref01_match_rt0, nil)
    assert.is_nil(err)
    assert.is_table(repo_ref01_list_rt0_result)

    local not_found_item = vs.select(
      runner.entity_list_to_data(repo_ref01_list_rt0_result),
      { id = repo_ref01_data["id"] })
    assert.is_true(vs.isempty(not_found_item))

  end)
end)

function repo_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/repo/RepoTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read repo test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "repo01", "repo02", "repo03", "owner01" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("GITHUB_TEST_REPO_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["GITHUB_TEST_REPO_ENTID"] = idmap,
    ["GITHUB_TEST_LIVE"] = "FALSE",
    ["GITHUB_TEST_EXPLAIN"] = "FALSE",
  })

  local idmap_resolved = helpers.to_map(
    env["GITHUB_TEST_REPO_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end
  if idmap_resolved["owner"] == nil then
    idmap_resolved["owner"] = idmap_resolved["owner01"]
  end

  if env["GITHUB_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["GITHUB_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["GITHUB_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end
